include/texture.hpp

```diff
@@ -33,74 +39,94 @@ struct checker_texture {
       : odd { solid_texture { c1 } }
       , even { solid_texture { c2 } } {}
   // Color value is different based on normalised spherical coordinates
-  color value(const hit_record& rec) const {
+  color value(const hit_record& rec, auto& ctx) const {
     auto sines =
         sin(10 * rec.p.x()) * sin(10 * rec.p.y()) * sin(10 * rec.p.z());
     if (sines < 0)
-      return odd.value(rec);
+      return odd.value(rec, ctx);
     else
-      return even.value(rec);
+      return even.value(rec, ctx);
   }
   solid_texture odd;
   solid_texture even;
 };
 
 /// A texture based on an image
 struct image_texture {
-
-  unsigned char* data {};
-  int width {};
-  int height {};
-
-  /// The repetition rate of the image
-  float cyclic_frequency { 1 };
-
-  int bytes_per_scanline {};
-
-  static constexpr auto bytes_per_pixel = 3;
-
   /** Create a texture from an image file
 
-        \param[in] file_name is the path name to the image file
+         \param[in] file_name is the path name to the image file
 
-        \param[in] cyclic_frequency is an optional repetition rate of
-        the image in the texture
-*/
-  image_texture(const char* file_name, float cyclic_frequency = 1)
-      : cyclic_frequency { cyclic_frequency } {
+         \param[in] cyclic_frequency is an optional repetition rate of
+         the image in the texture
+ */
+  static image_texture image_texture_factory(const char* file_name,
+                                             float _cyclic_frequency = 1) {
+    assert(!frozen);
     auto components_per_pixel = bytes_per_pixel;
-    data = stbi_load(file_name, &width, &height, &components_per_pixel,
-                     bytes_per_pixel);
-
-    if (!data) {
+    uint8_t* _data;
+    int _w, _h;
+    _data =
+        stbi_load(file_name, &_w, &_h, &components_per_pixel, bytes_per_pixel);
+    std::size_t _offset = 0;
+    if (!_data) {
       std::cerr << "ERROR: Could not load texture image file '" << file_name
                 << "'.\n"
                 << stbi_failure_reason() << std::endl;
-      width = height = 0;
+      _w = _h = 1;
+    } else {
+      auto size = bytes_per_pixel * _w * _h;
+      _offset = texture_data.size() / 3;
+      std::copy(_data, _data + size, std::back_inserter(texture_data));
     }
-    // \todo deallocate the image memory in the constructor
+    return image_texture(_w, _h, _offset, _cyclic_frequency);
+  }
 
-    bytes_per_scanline = bytes_per_pixel * width;
+  static sycl::buffer<uint8_t, 2> freeze() {
+    assert(!frozen);
+    frozen = true;
+    return sycl::buffer<uint8_t, 2> { texture_data.data(),
```

Comment:
why is this a 2D buffer ?
since images a chained one after the other, the 2D coordinate don't mean anything.